controllers/cm_event_handler.go

```diff
@@ -0,0 +1,59 @@
+package controllers
+
+import (
+	"context"
+
+	"github.com/openshift/sandboxed-containers-operator/internal/featuregates"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+type ConfigMapEventHandler struct {
+	reconciler *KataConfigOpenShiftReconciler
+}
+
+func (ch *ConfigMapEventHandler) Create(ctx context.Context, event event.CreateEvent, queue workqueue.RateLimitingInterface) {
+
+	if ch.reconciler.kataConfig == nil {
+		return
+	}
+
+	cm := event.Object
+
+	// Check if the configmap name is one of the feature gates configmap or
+	// feature config
+	if cm.GetNamespace() != OperatorNamespace || !featuregates.IsFeatureGateConfigMap(cm.GetName()) {
+		return
+	}
+	log := ch.reconciler.Log.WithName("CMCreate").WithValues("cm name", cm.GetName())
+	log.Info("FeatureGates configmap created")
+
+	queue.Add(ch.reconciler.makeReconcileRequest())
+}
+
+func (ch *ConfigMapEventHandler) Update(ctx context.Context, event event.UpdateEvent, queue workqueue.RateLimitingInterface) {
+
+	if ch.reconciler.kataConfig == nil {
+		return
+	}
+
+	cm := event.ObjectNew
+
+	// Check if the configmap name is one of the feature gates configmap or
+	// feature config
+	if cm.GetNamespace() != OperatorNamespace || !featuregates.IsFeatureGateConfigMap(cm.GetName()) {
+		return
+	}
+
+	log := ch.reconciler.Log.WithName("CMUpdate").WithValues("cm name", cm.GetName())
+	log.Info("FeatureGates configmap updated")
+
+	queue.Add(ch.reconciler.makeReconcileRequest())
+
+}
+
+func (ch *ConfigMapEventHandler) Delete(ctx context.Context, event event.DeleteEvent, queue workqueue.RateLimitingInterface) {
```

Comment:
If we reconcile on Create I guess it would make sense to also reconcile on Delete, wouldn't it?  I understand handling Create since we want to handle creation of both the feature gate master configmap and individual gated features' configmaps.  With the same reasoning though, wouldn't we want handle their deletion as well?

Reply:
How do you think the deletion of configmaps be handled?  Everything roll backed ? Or the configmap being recreated based on known state?
My thinking was to only handle create/update and document that on deletion the behaviour is not deterministic. And handle these cases when feature gate graduates to stable and part of KataConfig.

Reply:
Let me start by stating the wider context so that we can be sure we're on the same page.

I assume that we're only watching configmaps because we want to support modifications on-the-fly (ie. while a KataConfig exists).  If we didn't we'd just sample the configmaps during installation and then no further modification would be possible.

I believe it would then make sense to support all such modifications rather than a (pretty much arbitrary) subset.  I mean, we _could_ say "you can enable a gated feature subsequently if you didn't enable it initially but once you do that you're stuck with it, no disabling possible" but does it make much sense?

I'm not sure what you mean when you mention non-deterministic behaviour.  The feature gates are booleans so they fundamentally switch between two states, both of which necessarily need to be well defined.  If you delete a configmap that enabled both images and additional runtimeclasses I imagine that the controller should deterministically respond by switching back to extension and by deleting the extra runtimeclasses.

Reply:
> if you didn't enable it initially but once you do that you're stuck with it, no disabling possible" but does it make much sense?

What we say is you disable it by turning it off in the configmap and not by deleting the configmap.

I'm not comfortable to do a complete rollback for all extensions to default values if the configmap is deleted. IOW switching to extension, removing all runtimeClasses and other features we add going forward. 
Will be a heavy operation for say an accidental delete. Although one can argue that it's an admin operation and need to be done carefully.
Do you think this is how we should handle or there is an alternative?

Reply:
The question is, do we even have a choice?  Cluster admin can delete the configmap as they see fit.  Our only choice that I can see is, do we handle the deletion or do we allow the cluster to stay in an inconsistent state?

It's true that depending on the gated feature set a complete rollback might be a heavy operation and it could be done in error.  But the same is true of just about any KataConfig operation, including creation, modification and deletion.

Reply:
One approach I was thinking was to use the in-memory state of the featuregate to decide. But then it deviates from the using etcd as the source of truth. I'll do some experiments and update.

Reply:
In-memory state wouldn't be reliable enough, the controller could be restarted at any time.  And yes, it could be rather confusing for cluster admin, too, if the controller is guided by its hidden in-memory state rather the visible cluster resources...

Reply:
Adding a finalizer would indeed delay deletion, I'm afraid however that it would be quite a misuse (or even abuse) of finalizer.

Reply:
Yeah adding finalizer will not give us anything here imho..
Anyways I have added some changes to handle deletion.. PTAL